compiler-hacking.md

```diff
@@ -0,0 +1,464 @@
+this tutorial may be different from future output of the compiler. However, the
+core behaviour of the Shesmu compiler is based on Java's `Stream`, which is
+stable, so they will be functionally similar. Bytecode dumps can be seen on the
+olive dashboard.
+
+To start off, compiling a simple script:
+
+     Input shesmu;
+
+is compiled to:
+
+    public class dyn/shesmu/Program extends ca/on/oicr/gsi/shesmu/ActionGenerator {
+    
+      public <init>()V
+        ALOAD 0
+        INVOKESPECIAL ca/on/oicr/gsi/shesmu/ActionGenerator.<init> ()V
+        RETURN
+        MAXSTACK = 0
+        MAXLOCALS = 1
+    
+      private clearGauge()V
+        RETURN
+        MAXSTACK = 0
+        MAXLOCALS = 1
+    
+      public synchronized run(Lca/on/oicr/gsi/shesmu/ActionConsumer;Ljava/util/function/Function;)V
+        ALOAD 0
+        INVOKEVIRTUAL dyn/shesmu/Program.clearGauge ()V
+       L0
+       L1
+        RETURN
+        MAXSTACK = 0
+        MAXLOCALS = 3
+    
+      public static <clinit>()V
+        RETURN
+        MAXSTACK = 0
+        MAXLOCALS = 0
+    }
+
+This class, `dyn.shesmu.Program`, extends the `ActionGenerator` superclass,
+which is used by the server to execute actions. The class name is reused for
+every script, but since each is loaded by a separate class loader, they are not
+in conflict. The constructor is relatives boring; it just calls the
+super-constructor. The `run` method is where all the interesting things happen.
+The `clearGauge` method is executed at the start and performs all the necessary
+initialisation for gauges. This is a convenience of compiler design: the `run`
+and `clearGauge` methods can be built in parallel even though they are executed
+sequentially. Since this script has no olives, the `run` method does nothing
+else and returns.
+
+Let's move to a script with an olive:
+
+    Input shesmu;
+    Olive
+      Run nothing With value = type;
+
+
+The output bytecode is more complicated. The initialisation is much the same:
+    
+    public class dyn/shesmu/Program extends ca/on/oicr/gsi/shesmu/ActionGenerator {
+    
+      public <init>()V
+        ALOAD 0
+        INVOKESPECIAL ca/on/oicr/gsi/shesmu/ActionGenerator.<init> ()V
+        RETURN
+        MAXSTACK = 0
+        MAXLOCALS = 1
+    
+      private clearGauge()V
+        RETURN
+        MAXSTACK = 0
+        MAXLOCALS = 1
+    
+      public synchronized run(Lca/on/oicr/gsi/shesmu/ActionConsumer;Ljava/util/function/Function;)V
+        ALOAD 0
+        INVOKEVIRTUAL dyn/shesmu/Program.clearGauge ()V
+
+Now there are interesting differences. First, the current time is stored for timing the olive runtime.
+
+       L0
+        INVOKESTATIC java/lang/System.nanoTime ()J
+        LSTORE 3
+
+Now, a stream of the correct type must be started that will contain the right
+class. Since the supplied argument is `Function<Class<T>, Stream<T>>`, it's
+invoked with the class specified by the `Input` declaration:
```

Comment:
Why must a stream be created? 
How is the correct type determined? 
Where was that argument supplied, and how does it imply that 'it' (the stream? the argument?) should be invoked using the class from Input?

Reply:
I've rewritten this section.